src/query/functions/src/scalars/map.rs

```diff
@@ -244,4 +252,144 @@ pub fn register(registry: &mut FunctionRegistry) {
                     .any(|(k, _)| k == key)
             },
         );
+
+    registry.register_3_arg_core::<EmptyMapType, GenericType<0>, GenericType<1>, MapType<GenericType<0>, GenericType<1>>, _, _>(
+        "map_insert",
+        |_, _, insert_key_domain, insert_value_domain| {
+            FunctionDomain::Domain(Some((
+                insert_key_domain.clone(),
+                insert_value_domain.clone(),
+            )))
+        },
+        |_, key, value, ctx| {
+            let key_type = &ctx.generics[0];
+            if !key_type.is_boolean()
+                && !key_type.is_string()
+                && !key_type.is_numeric()
+                && !key_type.is_decimal()
+                && !key_type.is_date_or_date_time()
+            {
+                ctx.set_error(0, format!("map keys can not be {}", key_type));
+            }
+
+            let mut b = ArrayType::create_builder(1, ctx.generics);
+            b.put_item((key.into_scalar().unwrap(), value.into_scalar().unwrap()));
+            b.commit_row();
+            return Value::Scalar(MapType::build_scalar(b));
+        },
+    );
+
+    registry.register_3_arg_core::<NullableType<MapType<GenericType<0>, GenericType<1>>>, GenericType<0>, GenericType<1>, MapType<GenericType<0>, GenericType<1>>, _, _>(
+        "map_insert",
+        |_, source_domain, insert_key_domain, insert_value_domain| {
+            FunctionDomain::Domain(match source_domain.has_null {
+                true => Some((
+                    insert_key_domain.clone(),
+                    insert_value_domain.clone(),
```

Comment:
`insert_key_domain` and `insert_value_domain` need merge with the key and value domain in `source_domain`.